src/lib.rs

```diff
@@ -564,26 +564,33 @@ impl<S: Storage + Send + Sync + 'static> SelfEncryptor<S> {
         }
     }
 
-    /// Truncate the self_encryptor to the specified size (if extend, filled with 0u8).
-    pub fn truncate(&mut self, position: u64) -> bool {
+    /// Truncate the self_encryptor to the specified size (if extended, filled with 0u8).
+    //
+    // NOTE: Right now calls prepare_window with the full length of the truncated file
+    // because chunk sizes may change. Chunks that don't change don't actually need any
+    // treatment here so there's room for performance improvement, especially for large
+    // files.
+    pub fn truncate(&mut self, new_size: u64) -> bool {
+        if self.file_size == new_size {
+            return true;
+        }
         let old_size = self.file_size;
-        self.file_size = position;  //  All helper methods calculate from file size
-        if position < old_size {
-            self.sequencer.truncate(position as usize);
-            let last_chunk = self.get_chunk_number(position) + 1;
-            self.chunks.truncate(last_chunk as usize);
+        if new_size < old_size {
+            self.prepare_window(new_size, 0);   // call before changing self.file_size
+            self.file_size = new_size;
+            self.sequencer.truncate(new_size as usize);
+            let num_chunks = self.get_num_chunks(); // call with updated self.file_size
+            self.chunks.truncate(num_chunks as usize);
```

Comment:
@lgvz Agreed.  Would it be possible for you to contact me directly to discuss this further?  fraser.hutchison@maidsafe.net